// components/ColoredCalendar/coloredCalendar.js
import React from "react";
import moment from "moment";
import MonthPiker from "./monthPicker";
import PropTypes from "prop-types";
import "./style.css";

export default class ColoredCalendar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      activeRange: {from: null,to: null},
      activeMonths: props.range ?
        {left: props.range.from, right: props.range.to} :
        {left: props.initPicker.from, right: props.initPicker.to},
    };
  }
  
  getChildContext() {
    return {
      format: this.props.format,
      colors: this.props.colors,
      setActiveMonth: this.setActiveMonth,
      activeRange: this.state.activeRange,
      setRange: this.setRange,
      limit: this.props.limit,
    };
  }
  
  componentWillMount() {
    moment.updateLocale(this.props.lang, {
      week: {dom: this.props.firstDayOfWeek}
    });
  }
  
  setRange = (value) => {
    this.setState(prevState => {
      if(!this.props.soloChoose) {
        if (prevState.activeRange.from && prevState.activeRange.to) {
          prevState.activeRange.from = null;
          prevState.activeRange.to = null;
          prevState.activeRange.from ? prevState.activeRange.to = value : prevState.activeRange.from = value;
        } else {
          if (prevState.activeRange.from) {
            if (prevState.activeRange.from > value) {
              prevState.activeRange.to = prevState.activeRange.from;
              prevState.activeRange.from = value;
            } else prevState.activeRange.to = value;
          } else prevState.activeRange.from = value;
        }
      } else {
        prevState.activeRange.from = value;
        prevState.activeRange.to = value;
      }
      return prevState;
    }, () => this.props.switchDays(this.state.activeRange));
  };
  
  setActiveMonth = (type, date) => {
    this.setState(prevState => {
      prevState.activeMonths[type] = date;
      return prevState;
    }, () => this.props.switchMonth(type, date));
  };
  
  render() {
    return (
      <div className="Date-RangePicker">
        <MonthPiker
          key="left"
          type={"left"}
          date={this.props.initPicker.from}
          skipMonth={this.state.activeMonths.right}
        />
        <MonthPiker
          key="right"
          type={"right"}
          date={this.props.initPicker.to}
          skipMonth={this.state.activeMonths.left}
        />
      </div>
    );
  }
}

ColoredCalendar.defaultProps = {
  format: "YY.MM.DD",
  initPicker: {
    from: moment().format("YY.MM.DD"),
    to: moment().add(1, "month").format("YY.MM.DD"),
  },
  colors: [],
  lang: "en",
  limit: 0,
  firstDayOfWeek: 1,
  soloChoose: false,
  switchDays: () => {},
  switchMonth: () => {}
};

ColoredCalendar.propType = {
  firstDayOfWeek: PropTypes.number,
  colors: PropTypes.array,
  switchDays: PropTypes.func,
};

ColoredCalendar.childContextTypes = {
  format: PropTypes.string,
  colors: PropTypes.array,
  setActiveMonth: PropTypes.func,
  activeRange: PropTypes.object,
  setRange: PropTypes.func,
  limit: PropTypes.number,
};
