// components/ColoredCalendar/Day.js
import React from "react";
import PropTypes from "prop-types";
import moment from "moment/moment";
import { get } from "lodash";

export const Day = (props, context) => {
  const {day} = props;
  const {oneActiveDay, setRange, colors} = context;
  const active = get(context, 'actionRange.from') ?
    moment(day) < moment(context.activeRange.to) && moment(day) > moment(context.activeRange.from) ? "--active" :
      moment(day).format("DDMMYYYY") === moment(get(context, 'actionRange.to')).format("DDMMYYYY") ? "--end" :
        moment(day).format("DDMMYYYY") === moment(get(context, 'actionRange.from')).format("DDMMYYYY") ? "--start" : "" : "";
  const today = moment(day).format("DDMMYYYY") === moment().format("DDMMYYYY");
  const activeSolo = moment(oneActiveDay).format("DDMMYYYY") === moment(day).format("DDMMYYYY") && !today;
  const colorItem = colors && colors.length ? colors.find((color) => color.day === moment(day).format("DD") && color.month === moment(day).format("MM") && color.year === moment(day).format("YY")) : null;
  return (
    <div
      onClick={() => setRange(day)}
      style={activeSolo ? {background: "#4CC802"} : colorItem ? {background: colorItem.color} : null}
      className={`Date-Month__day Date-Month__day${active}${today ? " today" : ""}`}
    >{day.format("DD")}</div>
  );
};

export const DayShadow = (props) => {
  const {day} = props;
  const from = get(props, 'activeRange.from');
  const to = get(props, 'activeRange.to');
  const active = from ? moment(day) < moment(to) && moment(day) > moment(from) ? "--active" :
    moment(day).format("DDMMYYYY") === moment(to).format("DDMMYYYY") ? "--start" :
      moment(day).format("DDMMYYYY") === moment(from).format("DDMMYYYY") : "";
  return (
    <div
      className={`Date-Month__day Date-Month__day--shadow ${active}`}
    >{day.format("DD")}</div>
  );
};

Day.contextTypes = {
  activeRange: PropTypes.object,
  setRange: PropTypes.func,
  setColor: PropTypes.func,
  colors: PropTypes.array,
  oneActiveDay: PropTypes.object,
};

DayShadow.contextTypes = {
  activeRange: PropTypes.object,
};
